File: programmers/python/입문100/25.4_다음에올숫자/25-4.py
def solution(common):
    l = r = 0
    if common[0] - common[1] == common[1] - common[2]:
        l = common[1] - common[0]
    else:
        r = common[1] / common[0]
    # 차가 같으면 등차
    # 차가 같지 않으면 등비

    if not r:  # 등차수열이면(간격이 0일수도 있기 때문)
        return common[-1] + l
    elif r:  # 등비수열이면
        return common[-1] * r

File: programmers/java/기초124/25.4_이차원배열대각선순회/25_4.java
class Solution {
    public int[][] solution(int n) {
        // int[][] answer = {};
        int[][] answer = new int[n][n];

        // 오른쪽, 아래, 왼쪽, 위
        int[] dr = {0, 1, 0, -1};
        int[] dc = {1, 0, -1, 0};

        int r = 0, c = 0, dir = 0;
        for (int val = 1; val <= n * n; val++) {
            answer[r][c] = val;

            int nr = r + dr[dir];
            int nc = c + dc[dir];

            // 다음 칸이 범위 밖이거나 이미 값이 있으면 방향 전환
            if (nr < 0 || nr >= n || nc < 0 || nc >= n || answer[nr][nc] != 0) {
                dir = (dir + 1) % 4;
                nr = r + dr[dir];
                nc = c + dc[dir];
            }

            r = nr;
            c = nc;
        }
        return answer;
    }
}